cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh

```diff
@@ -0,0 +1,144 @@
+ * @param[in] linestring_offsets_end Iterator to the end of the range of the linestring offsets
+ * @param[in] linestring_points_begin Iterator to the begin of the range of the linestring points
+ * @param[in] linestring_points_end Iterator to the end of the range of the linestring points
+ * @param[out] distances Iterator to the output range of shortest distances between pairs.
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename Cart2dItA, typename Cart2dItB, typename OffsetIterator, typename OutputIterator>
+void __global__ pairwise_point_linestring_distance(Cart2dItA points_first,
+                                                   OffsetIterator linestring_offsets_first,
+                                                   OffsetIterator linestring_offsets_last,
+                                                   Cart2dItB linestring_points_first,
+                                                   Cart2dItB linestring_points_last,
+                                                   OutputIterator distances)
+{
+  using T = typename iterator_vec_base_type<Cart2dItA>;
+
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x;
+       idx < std::distance(linestring_points_first, linestring_points_last);
+       idx += gridDim.x * blockDim.x) {
+    auto offsets_iter =
+      thrust::upper_bound(thrust::seq, linestring_offsets_first, linestring_offsets_last, idx);
+    // Pointer to the last point in the linestring.
+    if (*offsets_iter - 1 == idx) { return; }
+
+    auto point_idx = thrust::distance(linestring_offsets_first, thrust::prev(offsets_iter));
+    cartesian_2d<T> const a = linestring_points_first[idx];
+    cartesian_2d<T> const b = linestring_points_first[idx + 1];
+    cartesian_2d<T> const c = points_first[point_idx];
+
+    auto const distance_squared = point_to_segment_distance_squared(c, a, b);
+
+    atomicMin(&thrust::raw_reference_cast(*(distances + point_idx)),
+              static_cast<T>(std::sqrt(distance_squared)));
+  }
+}
+
+}  // namespace detail
+
+template <class Cart2dItA, class Cart2dItB, class OffsetIterator, class OutputIt>
+void pairwise_point_linestring_distance(Cart2dItA points_first,
+                                        Cart2dItA points_last,
+                                        OffsetIterator linestring_offsets_first,
+                                        Cart2dItB linestring_points_first,
+                                        Cart2dItB linestring_points_last,
+                                        OutputIt distances_first,
+                                        rmm::cuda_stream_view stream)
+{
+  using T = typename iterator_vec_base_type<Cart2dItA>;
+
+  static_assert(detail::is_same_floating_point < T,
+                iterator_vec_base_type<Cart2dItB>,
+                iterator_value_type<OutputIt>,
```

Comment:
I actually think we have to be careful how we limit the output iterator. It's possible, for example, that the user wants to fuse a transform with the output using `thrust::make_transform_output_iterator()`.  In this case, will the `value_type` of that iterator be the same?  It might be a transform that accepts a floating point type, but converts it to something else...  We should still support that case.

I ran into this explicitly in `points_in_spatial_window`, where I needed to use a zip iterator for the output since the header-only API expects 2D points and the column-based API has separate x and y output columns.  Due to the transform output iterator used, I couldn't use the same type check on the output as on the input...

See https://github.com/rapidsai/cuspatial/blob/8dd2d9307259b741a7fcac3378606c2c7e69dde5/cpp/src/spatial_window/spatial_window.cu#L66-L70

Reply:
> In this case, will the `value_type` of that iterator be the same?

My original gut feeling says no, but the example says yes... https://godbolt.org/z/EfdPPY5re
... which left me confused because this seems inconsistent with standard. My reason is:

1. `transform_output_iterator` is an output iterator.
2. Output iterator's value_type in std must be `void`. According to https://en.cppreference.com/w/cpp/iterator/iterator_traits

Member type | Definition
-- | --
difference_type | std::incrementable_traits<Iter>::difference_type if valid, otherwise void
value_type | void
pointer | void
reference | void
iterator_category | std::output_iterator_tag

Am I missing something?

Reply:
Strictly speaking, there is no reason to constrain the output iterator type, right?

Like, if someone passes in an iterator that doesn't allow assignment from `float`, then it's going to fail to compile anyways.

Are you just hoping to provide a better error message?

Reply:
I guess we are just trying to set expectations, and fail reasonably, yes.

Reply:
I believe in one of the discussions we decided to relax the restriction on output iterator here. Although in reality the restriction is implied because `atomicMin` doesn't support transformed output iterators.